api/rest.go

```diff
@@ -4190,6 +4298,236 @@ func handleUpdatePublisherI18n(exec boil.Executor, id int64, i18ns []*models.Pub
 	return handleGetPublisher(exec, id)
 }
 
+//Labels
+func handleGetLabelList(cp utils.ContextProvider, exec boil.Executor, r *LabelsRequest) (*LabelsResponse, *HttpError) {
+	mods := []qm.QueryMod{
+		qm.InnerJoin("content_units cu on cu.id=content_unit_id and and cu.secure <= ?", allowedSecure(cp, common.PERM_LABEL_READ)),
+	}
+
+	// filters
+	if err := appendIDsFilterMods(&mods, r.IDsFilter); err != nil {
+		return nil, NewBadRequestError(err)
+	}
+	if err := appendDateRangeFilterMods(&mods, r.DateRangeFilter, "created_at"); err != nil {
+		return nil, NewBadRequestError(err)
+	}
+	//append approve state filter
+	if r.ApproveState.Valid {
+		mods = append(mods, qm.Where("approve_state = ?", r.ApproveState))
+	}
+
+	// count query
+	var total int64
+	countMods := append([]qm.QueryMod{qm.Select("count(DISTINCT id)")}, mods...)
+	err := models.Labels(exec, countMods...).QueryRow().Scan(&total)
+	if err != nil {
+		return nil, NewInternalError(err)
+	}
+	if total == 0 {
+		return NewLabelsResponse(), nil
+	}
+
+	labels, err := models.Labels(exec, qm.Load("LabelI18ns", "LabelI18ns.Author", "Tags")).All()
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, NewNotFoundError()
+		} else {
+			return nil, NewInternalError(err)
+		}
+	}
+
+	// i18n
+	data := make([]*Label, len(labels))
+	for i, pr := range labels {
+		x := &Label{Label: *pr}
+		data[i] = x
+		x.I18n = make(map[string]*LabelI18n, len(pr.R.LabelI18ns))
+		for _, i18n := range pr.R.LabelI18ns {
+			x.I18n[i18n.Language] = &LabelI18n{LabelI18n: *i18n, Author: i18n.R.User}
+		}
+	}
+
+	return &LabelsResponse{
+		ListResponse: ListResponse{Total: total},
+		Labels:       data,
+	}, nil
+}
+
+func handleCreateLabel(cp utils.ContextProvider, exec boil.Executor, r *CreateLabelRequest) (*models.Label, *HttpError) {
+	cu, err := models.ContentUnits(exec,
+		qm.WhereIn("uid = ?", r.ContentUnit),
+		qm.Where("secure <= ?", allowedSecure(cp, common.PERM_LABEL_WRITE)),
+	).One()
+	if err != nil {
+		return nil, NewInternalError(err)
```

Comment:
This 500 internal server error is confusing.
Permission errors are either 403 Forbidden or 404 Not found.

403 means: you are authenticated, I know who you are. I found the data you wish to modify. However you lack permission to do so.
404 means (in this case): I don't care who you are. I don't know what you're talking about. 

404 deliberately discloses less information

Reply:
fixed